src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ThrowHelper.cs

```diff
@@ -0,0 +1,97 @@
+
+namespace System.Text.RegularExpressions
+{
+    internal static class ThrowHelper
+    {
+        [DoesNotReturn]
+        internal static Exception CreateArgumentException(ExceptionResource resource) =>
+            throw new ArgumentException(GetStringForExceptionResource(resource));
+
+        [DoesNotReturn]
+        internal static void ThrowArgumentNullException(ExceptionArgument arg) =>
+            throw new ArgumentNullException(GetStringForExceptionArgument(arg));
+
+        [DoesNotReturn]
+        internal static void ThrowArgumentOutOfRangeException(ExceptionArgument arg) =>
+            throw new ArgumentOutOfRangeException(GetStringForExceptionArgument(arg));
+
+        [DoesNotReturn]
+        internal static void ThrowArgumentOutOfRangeException(ExceptionArgument arg, ExceptionResource resource) =>
+            throw new ArgumentOutOfRangeException(GetStringForExceptionArgument(arg), GetStringForExceptionResource(resource));
+
+        [DoesNotReturn]
+        internal static void ThrowNotSupportedException(ExceptionResource resource) =>
+            throw new NotSupportedException(GetStringForExceptionResource(resource));
+
+        private static string? GetStringForExceptionArgument(ExceptionArgument arg) =>
+            arg switch
+            {
+                ExceptionArgument.assemblyname => "assemblyname",
+                ExceptionArgument.array => "array",
+                ExceptionArgument.arrayIndex => "arrayIndex",
+                ExceptionArgument.count => "count",
+                ExceptionArgument.evaluator => "evaluator",
+                ExceptionArgument.i => "i",
+                ExceptionArgument.inner => "inner",
+                ExceptionArgument.input => "input",
+                ExceptionArgument.length => "length",
+                ExceptionArgument.matchTimeout => "matchTimeout",
+                ExceptionArgument.name => "name",
+                ExceptionArgument.options => "options",
+                ExceptionArgument.pattern => "pattern",
+                ExceptionArgument.regexinfos => "regexinfos",
+                ExceptionArgument.replacement => "replacement",
+                ExceptionArgument.startat => "startat",
+                ExceptionArgument.str => "str",
+                ExceptionArgument.value => "value",
+                _ => null
+            };
+
+        private static string? GetStringForExceptionResource(ExceptionResource resource) =>
+            resource switch
+            {
+                ExceptionResource.BeginIndexNotNegative => SR.BeginIndexNotNegative,
+                ExceptionResource.CountTooSmall => SR.CountTooSmall,
+                ExceptionResource.LengthNotNegative => SR.LengthNotNegative,
+                ExceptionResource.OnlyAllowedOnce => SR.OnlyAllowedOnce,
+                ExceptionResource.ReplacementError => SR.ReplacementError,
+                _ => null
+            };
+    }
+
+    internal enum ExceptionArgument
+    {
+        assemblyname,
```

Comment:
(nit) any reason why these enum values are camel cased?

Reply:
To exactly match the arguments they represent.  We do the same thing in other ThrowHelpers, e.g.
https://github.com/dotnet/runtime/blob/c26a65343448771e0dcdbcfdd5756c516a919e6b/src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs#L853-L942
https://github.com/dotnet/runtime/blob/c26a65343448771e0dcdbcfdd5756c516a919e6b/src/libraries/System.Memory/src/System/ThrowHelper.cs#L108-L124
https://github.com/dotnet/runtime/blob/c26a65343448771e0dcdbcfdd5756c516a919e6b/src/libraries/System.Linq/src/System/Linq/ThrowHelper.cs#L61-L80